### PR TITLE
Trace.raw mode for eager reading of traces

### PR DESCRIPTION
**Pull request by jokva**
Trace.raw mode for eager reading of traces

Adds support for the raw trace sub mode that does eager reading of
traces into an MxN numpy array - a useful feature for segyview and other
visualisation tools when it is known beforehand that the file will fit
in memory.

### CODE REVIEW

File: python/segyio/_raw_trace.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        """ :rtype: numpy.ndarray """
+        buf = None
+        if isinstance(index, slice):
+            f = self.trace._file
+            start, stop, step = index.indices(f.tracecount)
+            mstart, mstop = min(start, stop), max(start, stop)
+            length = max(0, (mstop - mstart + (step - (1 if step > 0 else -1))))
+            buf = np.zeros(shape = (length, f.samples), dtype = np.single)
+
+        return self.trace._readtr(index, buf)
```
Comment (jepebe) [inline python/segyio/_raw_trace.py]: What if index is int then but is None. How is this handled?

File: python/segyio/_segyio.c [hunk 2]
```diff
@@ -705,33 +705,62 @@ static PyObject *py_fread_trace0(PyObject *self, PyObject *args) {
 static PyObject *py_read_trace(PyObject *self, PyObject *args) {
     errno = 0;
     PyObject *file_capsule = NULL;
-    unsigned int trace_no;
+    PyObject *trace_no;
     PyObject *buffer_out;
+    Py_ssize_t start, stop, step, length;
```
Comment (jepebe) [inline python/segyio/_segyio.c]: Move!!!

File: python/segyio/segy.py [hunk 3]
```diff
@@ -302,6 +302,23 @@ def trace(self):
...
+            (especially memory-intensive) operations then this feature is
+            unlikely to help.
+
+            Read the first 10 traces::
+                >>> f.trace.raw[0:10]
+
+            Read *all* traces to memory::
+                >>> f.trace.raw[:]
+
+            Read every other trace traces to memory::
```
Comment (jepebe) [inline python/segyio/segy.py]: trace traces?

File: python/segyio/segy.py [hunk 4]
```diff
@@ -302,6 +302,23 @@ def trace(self):
             Fill the file with one trace (filled with zeros)::
                 >>> tr = np.zeros(f.samples)
                 >>> f.trace = itertools.repeat(tr)
+
+            For advanced users: sometimes you want to load the entire segy file
+            to memory and apply your own structural manipulations or operations
+            on it. Some segy files are very large and may not fit, in which
+            case this feature will break down. This is an optimisation feature.
+            If you're doing something trace-by-trace and performing non-trivial
```
Comment (jepebe) [inline python/segyio/segy.py]: I think the non-trivial memory intensive bit does not make sense. But maybe if you work trace-by-trace or line-by-line...

File: python/segyio/_raw_trace.py (hunk 1 above)
Comment (jokva) [inline python/segyio/_raw_trace.py]: By readtr which again calls _segyio.read_trace which soon also checks for None (both for slice and int).

**Comment by jokva** [comment]
Adress'd.

**Pull request merged**